gtfs/importers/gtfs_feed_importer.py

```diff
@@ -0,0 +1,279 @@
+        start_time = timer()
+
+        self.logger.info("Reading data...")
+        gtfs_feed = gtfs_kit.read_feed(url_or_filename, dist_units="km")
+
+        if not skip_validation:
+            self.logger.debug("Validating data...")
+            results = gtfs_kit.validate(gtfs_feed, as_df=False)
+            if results:
+                self.logger.info("Validation errors:")
+                self.logger.info(results)
+                return
+
+        with transaction.atomic():
+            # TODO just some temporary feed handling to get things rolling
+            feed, _ = Feed.objects.get_or_create(name=url_or_filename)
+
+            self._delete_existing_gtfs_objects(feed)
+
+            for model, gtfs_attribute in self.MODELS_AND_GTFS_KIT_ATTRIBUTES:
+                self._import_model(feed, model, getattr(gtfs_feed, gtfs_attribute))
+
+            # Update the feed's updated_at. We might want to do something else here.
+            feed.save()
+
+            self.logger.debug("Committing...")
+
+        end_time = timer()
+        self.logger.info(
+            f'Successfully imported a GTFS feed from "{url_or_filename}" in {end_time - start_time:.2f} secs!'
+        )
+
+    def _delete_existing_gtfs_objects(self, feed):
+        for model, _ in self.MODELS_AND_GTFS_KIT_ATTRIBUTES:
+            self.logger.debug(f"Deleting existing {model._meta.verbose_name_plural}...")
+            model.objects.filter(feed=feed).delete()
+
+    def _import_model(self, feed, model, gtfs_data):
+        num_of_rows = len(gtfs_data) if gtfs_data is not None else 0
+        plural_name = model._meta.verbose_name_plural
+
+        if num_of_rows:
+            self.logger.info(f"Importing {num_of_rows} {plural_name}...")
+        else:
+            self.logger.info(f"No {plural_name}")
+            return
+
+        objs_to_create = []
+        num_of_skipped = 0
+
+        for num_of_processed, gtfs_row in enumerate(gtfs_data.itertuples(), 1):
+            creation_attributes = {}
+
+            # Populate object creation attributes using the mapping from model fields
+            # to GTFS fields. How the values are converted between the two is
+            # determined by the model field's type.
+            for model_field_name, gtfs_field in self.FIELD_MAPPING[model].items():
+                gtfs_value = (
+                    getattr(gtfs_row, gtfs_field, None)
+                    if isinstance(gtfs_field, str)
+                    # when a model field is mapped to multiple GTFS fields build a
+                    # list of their values
+                    else [getattr(gtfs_row, f, None) for f in gtfs_field]
+                )
+                model_field = model._meta.get_field(model_field_name)
+                creation_attributes[model_field_name] = self._convert_value(
+                    gtfs_value, model_field, gtfs_field
+                )
+            if model == CalendarDate and creation_attributes["calendar_id"] is None:
```

Comment:
If this would be the last `CalendarDate` in the data we're importing and we skip it (`continue`), it might cause that we will not create up to 2000 `CalendarDate` instances since the last `bulk_create` might not get called. Right?

Reply:
:point_up: Removing `continue` and putting `objs_to_create.append(...` inside an `else`-clause might resolve that corner-case.

Reply:
Yes it would, good catch! 🎖️👍 I needed to add that stupid skipping there at the last minute, and apparently didn't really think it through properly.

Reply:
Fixed.